text/0066-separate-context-propagation.md

```diff
@@ -0,0 +1,499 @@
+single key. This allows the SDK to make improvements and leverage additional 
+data that may be available, without changes to all of the call sites.
+
+
+## Context Management and in-process propagation
+
+In order for Context to function, it must always remain bound to the execution 
+of code it represents. By default, this means that the programmer must pass a 
+Context down the call stack as a function parameter. However, many languages 
+provide automated context management facilities, such as thread locals. 
+OpenTelemetry should leverage these facilities when available, in order to 
+provide automatic context management.
+
+## Pre-existing Context implementations
+
+In some languages, a single, widely used Context implementation exists. In other 
+languages, there many be too many  implementations, or none at all. For example, 
+Go has a the `context.Context` object, and widespread conventions for how to 
+pass it down the call stack.
+
+In the cases where an extremely clear, pre-existing option is not available, 
+OpenTelemetry should provide its own Context implementation.
+
+## Default Propagators
+
+When available, OpenTelemetry defaults to propagating via HTTP header 
+definitions which have been standardized by the W3C.
+
+
+# FAQ
+
+## Why separate Baggage from Correlations?
+
+Since Baggage Context and Correlation Context appear very similar, why have two? 
+
+First and foremost, the intended uses for Baggage and Correlations are 
+completely different. Secondly, the propagation requirements diverge 
+significantly.
+
+Correlation values are solely to be used as labels for metrics and traces. By 
+making Correlation data write-only, how and when it is transmitted remains 
+undefined. This leaves the door open to optimizations, such as propagating the 
+correlation data out-of-band.
+
+Baggage values, on the other hand, are explicitly added in order to be accessed 
+by downstream by other application code. Therefore, Baggage Context must be 
+readable, and reliably propagated in-band in order to accomplish this goal.
+
+There may be cases where a key-value pair is propagated as a Correlation for 
+observability and as a Baggage item for application-specific use. AB testing is 
+one example of such use case. This would result in extra overhead, as the same 
+key-value pair would be present in two separate headers.  
+
+Solving this edge case is not worth having the semantic confusion of a single 
+implementation with a dual purpose.
+
+## What about complex propagation behavior?
+
+Some OpenTelemetry proposals have called for more complex propagation behavior. 
+For example, falling back to extracting B3 headers if W3C Trace-Context headers 
+are not found. Chained propagators and other complex behavior can be modeled as 
+implementation details behind the Propagator interface. Therefore, the 
+propagation system itself does not need to provide chained propagators or other 
```

Comment:
I see this argument, but I think the fact that this proposal includes a proposed API for chainedpropagators, and the fact that all practical examples would need chaining as propagating traces and baggage are critical for pretty much any tracing implementation, implies that chaining should be a top-level concept.

If there is no standard way to handle this now, I see two things happening:

1. OpenTelemetry implementations will fraction with their own behavior to solve multiple propagators. Any deviation between implementations decreases the transferable expertise of OpenTelemetry.
2.  propagators will not be designed keeping chaining in mind (e.g. checking to see if there already is a valid value for traces, or inserting a corrupted one regardless of the existing of any previous traces). 

I think at the very least individual propagator behavior should be designed on the understanding that chained propagation will happen. Without that we will hit a very nasty painpoint trying to compensate for invalid behavior outside of the propagator (such as creating needless additional contexts in the middle to safely validate new changes)

Reply:
Just to be clear, chaining is for combining propagators from separate systems. I don't think anyone should use the chaining to write a complex propagator for a single system.
```
// this is wrong
extractor = propagation.ChainHTTPExtractor(
              tracing.TraceContextExtractor(), 
              tracing.B3Extractor(),
              tracing.JaegerExtractor(),
              baggage.DefaultExtractor())

// this is correct
extractor = propagation.ChainHTTPExtractor(
              tracing.FallbackExtractor( "TraceContext", "B3", "Jaeger"),
              baggage.DefaultExtractor())
```

Does that clarify the intent?

Reply:
Yep! I think there should be a followup otep around the complex propagator, but it makes total sense not to hash it out in this OTEP.